### ISSUE TITLE
[TypeScript Resolvers] Apollo Server resolvers types missmatch

### ISSUE THREAD

**Issue by zelid** [dotansimha/graphql-code-generator#1133]
I use typed resolvers generation and define my resolvers like:

```
import { QueryResolvers, MutationResolvers, SubscriptionResolvers, User } from './generated/graphql'

const query: QueryResolvers.Resolvers = {
  allUsers: (root, args, ctx) => {
    const result: User[] = []
    return result
  }
}

const mutation: MutationResolvers.Resolvers = {
  signup: (root, args, ctx) => {
    const res = ctx.gql.mutation.createUser({
      input: {
        user: {
          firstName: 'John Doe'
        }
      }
    })
    const result: User = { firstname: 'asdf', lastname: 'asdfasdf' }
    return result
  }
}

export default {
  mutation,
  query
}

```

When it comes to configure apollo-server like this:

```
import resolvers from './app/resolvers'

const server = new ApolloServer({
      typeDefs,
      resolvers,
      mocks: true,
      context: (c: any) => ({
        ...c,
        orm: connection,
        gql: postgraphileConnection
      })
    })
```

then apollo-server gives an error about types mismatch:

```
Type '{ mutation: Resolvers<GqlContext, {}>; query: Resolvers<GqlContext, {}>; }' is not assignable to type 'IResolvers<any, any>'.
  Property 'mutation' is incompatible with index signature.
    Type 'Resolvers<GqlContext, {}>' is not assignable to type '(() => any) | GraphQLScalarType | IEnumResolver | IResolverObject<any, any> | IResolverOptions<any, any>'.
      Type 'Resolvers<GqlContext, {}>' is not assignable to type 'IResolverObject<any, any>'.
        Index signature is missing in type 'Resolvers<GqlContext, {}>'.ts(2322)

types.d.ts(28, 5): The expected type comes from property 'resolvers' which is declared here on type 'Config & { cors?: boolean | CorsOptions; }'
```

What should be done in order to solve it?

**Comment by ardatan** [comment]
Could you use title case for Query and Mutation resolvers instead of query and mutation?

**Comment by zelid** [comment]
I've tried that as well, the error stays the same:

```
Type '{ Mutation: Resolvers<GqlContext, {}>; Query: Resolvers<GqlContext, {}>; }' is not assignable to type 'IResolvers<any, any>'.
  Property 'Mutation' is incompatible with index signature.
    Type 'Resolvers<GqlContext, {}>' is not assignable to type '(() => any) | GraphQLScalarType | IEnumResolver | IResolverObject<any, any> | IResolverOptions<any, any>'.
      Type 'Resolvers<GqlContext, {}>' is not assignable to type 'IResolverObject<any, any>'.
        Index signature is missing in type 'Resolvers<GqlContext, {}>'.ts(2322)
types.d.ts(28, 5): The expected type comes from property 'resolvers' which is declared here on type 'Config & { cors?: boolean | CorsOptions; }'
```

**Comment by mariusmarais** [comment]
As I understand it, the reason for this is because Apollo's `IResolvers` is defined as a dictionary / with an index signature (see [the code](https://github.com/apollographql/graphql-tools/blob/2394f0c3a8b30dbd4435dd21019727e76e61b892/src/Interfaces.ts#L96)).

This is not the same as an object with keys, even if the keys would match the index type.  TS 2.0 was supposed to fix this (I think?) but it still doesn't work for me.

I would do the following:
```ts
import { IResolvers } from 'graphql-tools';
import resolvers from './app/resolvers'

const server = new ApolloServer({
      typeDefs,
      resolvers as IResolvers, // <------- HERE
      mocks: true,
      context: (c: any) => ({
        ...c,
        orm: connection,
        gql: postgraphileConnection
      })
    })
```
Although it doesn't really fix it.

(FYI, I also have `Query` and `Mutation`, etc., and not lower-case.)

**Comment by zelid** [comment]
Thanks @mariusmarais, this helped:

```
const resolvers: IResolvers = {
  Query: {
    ...Query
  },
  Mutation: {
    ...(Mutation as IResolvers)
  }
}
```

**Comment by kamilkisiela** [comment]
I can confirm it's the issue @mariusmarais mentioned. I don't think we can do anything about it right now, without implementing some weird workarounds. It's something related to TypeScript, not GraphQL Code Generator, at least for now.

**Comment by vadistic** [comment]
Hello everyone, I think I can help :smile: 

As @mariusmarais said Apollo resolvers are defined as string indexed dictionaries. __Which is good__ because it allows dynamic propery access in strict mode like `resolvers[varableName]` and exclude otherwise possible number/symbols indexes.

There are few ways to provide string index signatures, my usual:
```ts
interface StringIndexSignatureInterface {
  [index: string]: any
}

type StringIndexed<T> = T & StringIndexSignatureInterface
```

And now you can do:
```ts
// no issues with apollo-server & no type assertion necessary
const resolvers: StringIndexed<IResolvers> = {...}
```

If `graphql-code-generator-typescript-server` is meant **only** for Apollo - this could be added to generated code. Otherwise - I think everything is ok (because removing index signatures is somehow harder than adding them^^) -  maybe sb can mention of this 'trick' in docs.

**Comment by noahseger** [comment]
@vadistic doesn't this approach remove all type information from the resolver definitions? I don't see a meaningful difference between that and `const resolvers: any = { /* ... */ }`.

**Comment by vadistic** [comment]
@noahseger No - Au Contraire :smirk: 

We're not removing typing because it's type intersection. Prop value needs to be `any & whateverwasbefore = whateverwasbefore`.

[Lazy Example](https://www.typescriptlang.org/play/index.html#src=const%20obj%20%3D%20%7B%0D%0A%20%20%20%20someProp%3A%20'abc'%2C%0D%0A%20%20%20%20someOtherProp%3A%201%0D%0A%7D%0D%0A%0D%0Atype%20Obj%20%3D%20typeof%20obj%0D%0A%0D%0Ainterface%20StringIndexSignatureInterface%20%7B%0D%0A%20%20%5Bindex%3A%20string%5D%3A%20never%0D%0A%7D%0D%0A%0D%0Atype%20StringIndexed<T>%20%3D%20T%20%26%20StringIndexSignatureInterface%0D%0A%0D%0Atype%20Test%20%3D%20StringIndexed<Obj>%0D%0A%0D%0Atype%20Valid%20%3D%20Test%5B'someOtherProp'%5D%0D%0A%0D%0Aconst%20indexedObj%20%3D%20obj%20as%20StringIndexed<Obj>%0D%0A%0D%0Aconst%20test%20%3D%20indexedObj%5B'someOtherProp'%5D%0D%0A%0D%0A%2F%2F%20Losing%20only%20this%20this%0D%0Aconst%20test2%20%3D%20indexedObj%5B'a'%5D%0D%0Aconst%20test3%20%3D%20obj%5B'a'%5D%0D%0A%0D%0A%2F%2F%20but%20not%20really%0D%0AindexedObj%5B'a'%5D%20%3D%201%0D%0AindexedObj%5B'a'%5D%20%3D%20'abc'%0D%0A%0D%0Atype%20AltIndexed<T%2C%20K%20extends%20string>%20%3D%20%7B%0D%0A%20%20%20%20%5BK%20in%20keyof%20T%5D%3A%20T%5BK%5D%0D%0A%7D%20%0D%0A%0D%0Aconst%20altIndexedObj%20%3D%20obj%20as%20AltIndexed<Obj%20%2Cstring>%0D%0A%0D%0A%2F%2F%20is%20indexed%0D%0Aconst%20test4%20%3D%20altIndexedObj%5B'someOtherProp'%5D%0D%0A%2F%2F%20disallowed%0D%0Aconst%20test5%20%3D%20altIndexedObj%5B'a'%5D%0D%0A%0D%0A) <= turn all the strict in Options!

The only thing lost with it is allowing accessing other typed indexes and receiving proper undefined type. (which is good thing for me usually - reflecting js, allowing lookups of props - and that's why I proposed it)

It could be tightened a bit with this:
```ts
// You can still access other (non-typed) indexes but you cannot assign anything to them
interface StringIndexSignatureInterface {
  [index: string]: never
}

type StringIndexed<T> = T & StringIndexSignatureInterface
```

Or there is another method using mapped type that is the strictest possible:
```ts
// Nothings lost, nothings added except string  indexes
type AltIndexed<T, K extends string> = {
    [K in keyof T]: T[K]
} 

const altIndexedObj = obj as AltIndexed<Obj ,string>
```

It all works, check in my lazy example :upside_down_face:

**Comment by dotansimha** [comment]
Fixed in 0.18.0 🎉

**Comment by panzelva** [comment]
Hello, I am not sure if relevant to this issue, but I am getting this error:

![I50M6A7U6s](https://user-images.githubusercontent.com/28455278/54266773-57075080-4578-11e9-80f7-a38f2ac0a8ce.png)

I am using `typescript@3.3.3333` and `graphql-code-generator@0.18.0`

**Comment by dotansimha** [comment]
@panzelva it was fixed in the recent refactor, and it will be available soon in 1.0.0.

**Comment by P4sca1** [comment]
I am still getting this issue with `v1.0.0`.

My resolvers:
```ts
import { IResolvers } from '../generated/graphql'
import { GameServer } from './GameServer'
import { Mutation } from './Mutation'
import { Query } from './Query'
import { User } from './User'

const resolvers: IResolvers = {
  GameServer,
  Mutation,
  Query,
  User,
}

export default resolvers
```

```ts
export interface IApolloContext {
  prisma: Prisma
}

const apolloServer = new ApolloServer({
  context: {
    prisma: new Prisma(),
  },
  resolvers,
  typeDefs: parse(importSchema(path.join(__dirname, '../schema.graphql'))),
})
```

Error:

> (property) Config.resolvers?: IResolvers<any, any> | undefined
Type 'IResolvers<IApolloContext>' is not assignable to type 'IResolvers<any, any>'.
  Property 'GameServer' is incompatible with index signature.
    Type 'GameServerResolvers<IApolloContext, GameServer> | undefined' is not assignable to type '(() => any) | GraphQLScalarType | IEnumResolver | IResolverObject<any, any, any> | IResolverOptions<any, any, any>'.
      Type 'undefined' is not assignable to type '(() => any) | GraphQLScalarType | IEnumResolver | IResolverObject<any, any, any> | IResolverOptions<any, any, any>'.ts(2322)
types.d.ts(30, 5): The expected type comes from property 'resolvers' which is declared here on type 'Config'

**Comment by kamilkisiela** [comment]
It's hidden behind a flag because you type safety is less strict then.

Try `useIndexSignatures: true`

**Comment by P4sca1** [comment]
That's it! No more problems using `useIndexSignature: true` (notice that there is no 's' at the end).

**Comment by aaronjensen** [comment]
@kamilkisiela would you mind editing your comment to remove the `s`? (and thanks @P4sca1 )

**Comment by ankitjena** [comment]
I am getting `Index signature is missing in type 'Resolvers'`. I tried using `useIndexSignature: true` after reading through the comments, but the problem persists.

```
TSError: ⨯ Unable to compile TypeScript:
src/index.ts:29:5 - error TS2322: Type 'Resolvers' is not assignable to type 'IResolvers<any, any> | IResolvers<any, any>[] | undefined'.
  Type 'Resolvers' is not assignable to type 'IResolvers<any, any>'.
    Index signature is missing in type 'Resolvers'.
```
I am using programmatic usage. What should I do?

**Comment by dotansimha** [comment]
@ankitjena which version do you use? 
It seems like your configuration is valid, the `useIndexSignature: true` should make it compatible with Apollo server.

**Comment by flybayer** [comment]
For anyone else who lands here and `useIndexSignature: true` isn't working — you have to kill the codegen server and restart it. The config hot reload doesn't work for this.

**Comment by esemeniuc** [comment]
For anyone wondering where to put the `useIndexSignature: true`, my codegen.yml looks like this (see the last 2 lines):

```yml
overwrite: true
schema: "src/schema.graphql"
documents: null
generates:
  src/generated/graphql.ts:
    plugins:
      - "typescript"
      - "typescript-resolvers"
    config:
      useIndexSignature: true
```

**Comment by dzmitry-kankalovich** [comment]
I literally spent an entire day on this problem. I wish things like this would be mentioned in README, something about known issues.

**Comment by Urigo** [comment]
@dzmitry-kankalovich would you mind having a go at a PR that would make it easier for the next to follow?

**Comment by dotansimha** [comment]
@dzmitry-kankalovich It's documented in our docs website, you can even find it easily in the search: 
![image](https://user-images.githubusercontent.com/3680083/82218922-49401e80-9925-11ea-894b-e454f0bc4bc0.png)

And there is a section for integration with Apollo Server: https://graphql-code-generator.com/docs/plugins/typescript-resolvers#integration-with-apollo-server

This is not a "known issue", because the resolver signature is agnostic to any GraphQL platform. If you wish to have it running with Apollo Server, you'll need to add the flag, because Apollo-Server expects a different signature than plain `graphql.js`.

**Comment by dzmitry-kankalovich** [comment]
@dotansimha You are absolutely right, it's there, and I've overlooked it. I am a newbie to TS and the `index signature` keywords were not catching my attention TBH.

I guess my larger problem was that I falsely expected it to "just work", considering that JS + Apollo is probably the most common starting point for a new GraphQL development - which is not really a problem of the codegen, but rather of Apollo.

I may suggest dropping a code snippet to that section, so it's just a bit more eye-catchy. Or maybe rephrase the doc like `By default Apollo Server will not work with generated Resolver interface...`.

Thanks!

**Comment by JustFly1984** [comment]
@dzmitry-kankalovich well, it definitely fixed an issue for me, but this hack kills type safety...

**Comment by dotansimha** [comment]
> @dzmitry-kankalovich well, it definitely fixed an issue for me, but this hack highlights incorrect types for ApolloServer resolvers.

Can you please share more details? The main issue here is the fact that Apollo-server signature expects a general type with index signature, while codegen can generate the exact named types and fields based on the schema. Then, a mismatch happens because the type provided by codegen does not match exactly to the open signature of Apollo-Server. 

As suggested above, using `useIndexSignature: true, noSchemaStitching: true` will make sure to create slim TS signature (without schema stitching resolvers) and with the index signature that Apollo-Server expects, and this should work without casting to `any`. 

If that doesn't work for you, can you please report a new issue with a reproduction and an example? Thank you!

**Comment by JustFly1984** [comment]
@dotansimha I have generated resolvers types as recommended, and now I have diffrent errors in resolvers.

Basically in some cases types forcing me to make resolvers of dictionary/Record/object type, where each field is resolver `async (parent, args, context, info) => value`, and errors out if I want to use `async function (args, context, info)` directly.

In other cases it accepts `async function` without errors, and I have no idea why.

Originally schema.json is generated from old Rubi graphql api.

**Comment by ardatan** [comment]
@JustFly1984 `async function (args, context, info)` is not a valid resolver signature for GraphQL-js. I didn't understand the use-case. Could you give more details, maybe an example CodeSandbox in a different issue?

**Comment by JustFly1984** [comment]
@ardatan basically I have next query structure:

```
query {
  page(id: ID) {
    id
    type
    title
 	item(id: ID) {
 		id
		content
        type
 	}
  }
}
```

resolver page:
```ts
import { item as itemResolver } from './item-resolver'

export async function page(parent: void, args, context, info): Promise<Page> {
  const { id, type, title } = await context.models.Pages....
  return {
    id,
    type,
    title,
    // Here it complains on function, and requires plain object
  	item: itemResolver,
  }
}
```

item-resolver:
```ts
export async function item(parent: Page, context, info): Promise<Item> {
  const { id, type, content } = await context.models.Items....
  return { id, type, content }
}
```

resolvers:
```ts
import { page } from './page'
import { item } from './item'

export const resolvers: Resolvers<ResolversContext> = {
  Page: page,
  // Here it complains on function, 
  // and requires an object with some props as resolver function type 
  // { 
  //.   id: (parent: Item, context, info) => id, 
  //.   type: (parent: Item, context, info) => 'Item',
  //.   content: 'Item Content'
  //  }
  // and others simple types.
  Item: item,
```

I do not understand why parent type is Item, and how should I get Args as it is described in query type.

**Comment by ardatan** [comment]
As I asked above, if anyone has an issue, please create a new issue with a clear minimal production on CodeSandbox. I'm locking this issue because this issue is pretty old.
@JustFly1984 I don't think resolvers can be like that;
https://www.graphql-tools.com/docs/resolvers
